[DTensor] Add Dijkstra-based single-dim strategy search

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #176221
* #175999
* __->__ #169438

Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

wconstab committed:
[DTensor] _find_lowest_cost_sharding

very WIP

[ghstack-poisoned]

Skylion007 commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -62,6 +69,476 @@ def spec_to_strategy(spec: object) -> object:
+            )
+            return OpStrategy([op_spec])
+
+        def can_chunk(placements: tuple[Placement, ...]) -> bool:
```
These also should return TypeIs with Tuple args

wconstab committed:
Update on "[DTensor] _find_lowest_cost_sharding"


very WIP

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] _find_lowest_cost_sharding"


very WIP

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] _find_lowest_cost_sharding"


very WIP

[ghstack-poisoned]

pianpwk commented on torch/distributed/tensor/_ops/single_dim_strategy.py:
```diff
@@ -214,3 +221,269 @@ def expanded_strategy(
+                            heapq.heappush(
+                                pq,
+                                (
+                                    new_total_cost,
```
I think using some notion of "breadth"/# of redistributes required helps here, otherwise for ops like pointwise, we might now unnecessarily chunk as it's no cost difference (e.g. `RR + S(0)R -> S(0)R` vs. `RR + S(0)R -> S(0)S(0)`)?

I haven't verified but just `(new_total_cost, n_comms + 1)`?

wconstab replied:
i definitely want to keep pointwise decoupled from this logic.  I would like to solve each group of inputs+output separately with caching.  So if we have 1k foreach inputs and likely some similar input placements patterns, we just do the graph search for each unique pattern once

pianpwk committed:
Update

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] _find_lowest_cost_sharding"


very WIP

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] _find_lowest_cost_sharding"


very WIP

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] _dijkstra_expand_single_dim_strategy_to_mesh"


very WIP

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]

zpcore reviewed (commented):
My current understanding is that when do graph search, the code will change one placement at a time as in `_get_neighbor_placements`. The thing is that it didn't consider the device order.
E.g., for 2D tensor matmul on 2D mesh, if we begin with lhs=(S(0), R), rhs=(R, S(0)), which doesn't fit into a valid matmul single dim strategy. So we will find neighbors. One neighbor for rhs is (S(0), S(0)), but this should be invalid because we can't go (R, S(0)) --> (S(0), S(0)) with one collective. I think we can add a test by only using the `transitions` to redistribute the input args and do the matmul, instead of calling redistribute separately after strategy generation. We may see numerical error of the matmul result due to incorrect distribution.

I think the transition should be: try all possible collective/chunk op on the current state and we can get the next state, then we check if the next state fits a valid single dim strategy.

wconstab committed:
Update on "[DTensor] Add Dijkstra-based single-dim strategy search"


Add _dijkstra_expand_single_dim_strategy_to_mesh, a priority-queue search
over input placement states that finds the lowest-cost sharding for an op
without enumerating all S^N strategy combinations (S = single-dim strategies,
N = mesh dimensions). The search uses _PreparedSingleDimStrategy (from the
previous commit) to materialize single-dim rules and try_propagate() to
test whether a candidate state matches on every mesh dimension.

Each search state is a tuple of per-input placement tuples. Neighbors are
generated by changing one placement on one mesh dimension for one input,
using _get_neighbor_placements which encodes DTensor redistribute transition
rules (Replicate <-> Shard, Partial -> Replicate/Shard). Cost computation
uses _compute_redistribute_cost which calls _compute_placement_transition_cost
directly per mesh dimension, avoiding DTensorSpec construction and
_gen_transform_infos planning overhead.

Returns None for _StridedShard inputs, signaling the caller to fall back to
full expansion.

Authored with Claude.

[ghstack-poisoned]
